From 3d2ddcaa1962ece5cccb5979c7f4c1992682284a Mon Sep 17 00:00:00 2001
From: Ilia Bozhinov <ammen99@gmail.com>
Date: Thu, 21 Mar 2024 17:34:37 +0200
Subject: [PATCH] output-layout: select mode with closest matching refresh rate
 (#2229)

This makes it possible to avoid the annoying cases where the monitor
provides 59994 as a refresh rate, and the user cannot say to use @60
instead.
---
 src/core/output-layout.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/core/output-layout.cpp b/src/core/output-layout.cpp
index 054b57789..d49d8ac5e 100644
--- a/src/core/output-layout.cpp
+++ b/src/core/output-layout.cpp
@@ -67,7 +67,9 @@ wlr_output_mode *find_matching_mode(wlr_output *output,
                 return mode;
             }
 
-            if (!best || (best->refresh < mode->refresh))
+            const int bestSoFar = std::abs(best->refresh - reference.refresh);
+            const int current   = std::abs(mode->refresh - reference.refresh);
+            if (!best || (bestSoFar > current))
             {
                 best = mode;
             }
